### DISCUSSION TITLE
Vale DITA-OT plugin — how to see the output in stdout?

### DISCUSSION THREAD

**Discussion by theletterf** [dita-ot/dita-ot#4484] [Q&A] [answered]
Hello there!

I'm trying to build a DITA-OT plugin for [Vale](https://vale.sh). I'm quite new to Ant and DITA-OT plugins, so forgive me beforehand.

This is my plugin.xml file:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://www.dita-ot.org/rng/plugin.rnc" type="application/relax-ng-compact-syntax"?>
<plugin id="org.splunk.vale.check" version="1.0">
   <feature extension="package.support.name" value="Fabrizio Ferri Benedetti"/>
   <feature extension="package.support.email" value="fferribenedetti@splunk.com"/>
   <feature extension="package.version" value="1.0"/>
   <transtype name="vale" desc="Check against style guide"/>
   <feature extension="ant.import" file="vale.xml"/>
</plugin>
```

And this is the vale.xml file:

```xml 
<?xml version='1.0' encoding='UTF-8'?>
<project name="dita2vale" default="vale-check" basedir="." xmlns:artifact="antlib:org.apache.maven.artifact.ant">
  <target name="dita2vale">
      <exec executable="vale">
          <arg value="> output.log" />
      </exec>
  </target>
</project>
```

The plugin installs with no error, but doesn't show any output to stdout (`<exec>` should show it by default if I'm not mistaken). At this point, I'm just trying to get *any* output to stdout. Then again, I'm not even sure if the approach I'm following is adequate — the idea here is to pass a ditamap to DITA OT and let the plugin output style issues based on system-wide Vale settings.

I'm open to suggestions. Thanks in advance for your patience.

**Comment by raducoravu** [comment]
Could you paste the command line that you are running to test the plugin? Have you tried adding the "-v" flag?
Inside your "dita2vale" can you add an Ant <echo> element with a message to see if that is visible?
Does the vale parser allow parsing directly the DITA XML content or do you plan to convert it to plain HTML and parse that?
As a sample a more complex DITA OT plugin which applies Schematron on DITA XML content: https://github.com/stefan-jung/org.jung.schematron

**Reply by theletterf** [reply]
Not currently at that computer, but if I recall correctly is `dita map.ditamap --format=vale` or something very similar (I'll confirm). I also tried verbose mode, to no avail. `<echo>` tasks are indeed visible when I add them.

At the very least, I'd expect `vale` to output an error, but it doesn't. When I pass `ls` as the command, nothing happens either, whch makes me think that something might be amiss in my plugin.

`vale` does indeed parse DITA files provided `dita` is in the system path, though conversion to HTML could also be an option.

**Reply by raducoravu** [reply]
How about if you remove this part? 

    <arg value="> output.log" />

On what operating system are you running the build?

**Reply by theletterf** [reply]
Running on the latest macOS. Not getting any output when I remove that line. Changing to useing `depend.preprocess.post` also doesn't yield results.

**Reply by raducoravu** [reply]
How about if you set failonerror=true on the "exec" command? Maybe you could also zip the entire plugin, attach it here and I could take a look at it.

**Reply by theletterf** [reply]
Here's the ZIP file! I run the `rerun.sh` script to test the plugin each time.

[ditasample.zip](https://github.com/user-attachments/files/15870993/ditasample.zip)

**Reply by theletterf** [reply]
Just validated that the following works when executing it directly with Ant:

```
<?xml version='1.0' encoding='UTF-8'?>
<project name="vale-check" default="vale-check" basedir="." xmlns:artifact="antlib:org.apache.maven.artifact.ant">
  <target name="vale-check">
      <echo message="I'm here!" />
      <exec executable="/usr/local/bin/vale" failonerror="true">
         <arg value="-v" />
         <redirector output="vale_output.log" error="vale_error.log"/>
      </exec>
      <echo message="I'm there!" />
  </target>
</project>
```

So it's a DITA OT issue now. Echo statements are executed when trying the DITA OT plugin, but no exec seems to happen.

**Reply by raducoravu** [reply]
It's good that you removed that spawn="true" attribute you had on the "exec", according to the docs:
   If you spawn a command, its output will not be logged by Ant.
https://ant.apache.org/manual/Tasks/exec.html

If I modify the build.xml to look something like this:

    <target name="vale-check">
      <exec executable="vale" spawn="false" logError="true">
      </exec>
    </target>

and I have already installed "vale" using homebrew, I get the following error running the DITA OT:

     sh dita -i /Users/raducoravu/eclipse-workspace/eXml/samples/dita/flowers/flowers.ditamap -f html5
     [exec] E100 [.vale.ini not found] Runtime error
     [exec] 
     [exec] no config file found
     [exec] 
     [exec] Execution stopped with code 1.
     [exec] Result: 2

 so this seems to work, vale is invoked.

**Reply by theletterf** [reply]
Thanks! We're closing in. I still don't get a Vale output when running `dita --input=sample.ditamap --format=html5`, which makes me think there's a macOS issue at play here. When I run `dita --input=sample.ditamap --format=html5 --verbose` I see the step is indeed executed (in this case, `vale` with `-v` argument):

```
==> Preprocess post
vale version 3.4.2
```

The fact that the same output doesn't show to stdout using zsh on macOS puzzles me deeply.

**Reply by theletterf** [reply, marked as answer]
I solved this by doing this:

```
      <exec executable="vale" failonerror="true" spawn="false" logError="true" outputproperty="cmdOutput" errorproperty="cmdError">
      </exec>
      <!-- Print the standard output to the console -->
      <echo>Standard Output:</echo>
      <echo>${cmdOutput}</echo>

      <!-- Print the error output to the console, if any -->
      <echo>Standard Error:</echo>
      <echo>${cmdError}</echo>
```

Now I only have to figure out how to pass each DITA file preprocessing path to Vale... :)

**Comment by infotexture** [comment]
👋 Hi Fabrizio, glad to see you're exploring this, could be very useful to the community.

A few years ago, we considered running Vale on the DITA-OT docs, and I opened a pull request to test it: https://github.com/dita-ot/docs/pull/284.  
(We ran into a few issues, and never ended up merging that.) 

IIRC, it had to do with performance issues that were related to the way Vale parsed DITA content at the time. I think back then, Vale would run the toolkit process on each DITA file, convert it to HTML, and run its checks on that. But that was a long time ago, and both Vale and DITA-OT have changed since, so it may work differently now.

Anyhow, not sure how relevant that old PR will be for your efforts here, but thought I should mention it in case there's anything there that's useful to you.

**Reply by theletterf** [reply]
Thanks Roger, I'll take a look! Glad to become part of the DITA OT community. :)
